includes/class-wc-google-gtag-js.php

```diff
@@ -49,13 +49,40 @@ public function __construct( $settings = array() ) {
 		$this->map_hooks();
 
 		// Setup frontend scripts
+		add_action( 'wp_head', array( $this, 'setup_site_tag' ), 2 );
 		add_action( 'wp_enqueue_scripts', array( $this, 'enquque_tracker' ), 5 );
 		add_action( 'wp_footer', array( $this, 'inline_script_data' ) );
 	}
 
 	/**
-	 * Register tracker scripts and its inline config.
-	 * We need to execute tracker.js w/ `gtag` configuration before any trackable action may happen.
+	 * Setup the global site tag as early as possible on the page
+	 *
+	 * @return void
+	 */
+	public function setup_site_tag() {
+		wp_print_inline_script_tag(
+			sprintf(
+				'/** Google Analytics for WooCommerce (gtag.js) */
+				window.dataLayer = window.dataLayer || [];
+				function %2$s(){dataLayer.push(arguments);}
+				// Set up default consent state.
+				for ( const mode of %4$s || [] ) {
+					%2$s( "consent", "default", mode );
+				}
+				%2$s("js", new Date());
+				%2$s("set", "developer_id.%3$s", true);
+				%2$s("config", "%1$s", %5$s);',
+				esc_js( self::get( 'ga_id' ) ),
+				esc_js( self::tracker_function_name() ),
+				esc_js( self::DEVELOPER_ID ),
+				json_encode( self::get_consent_modes() ),
+				json_encode( $this->get_site_tag_config() )
```

Comment:
Are we tied to legacy function names / visibility here? Just looks confusing to be mixing `self` and `$this`.

Reply:
I had a bigger concern with the `self::` functions, especially `self::get`. 
But I didn't want to push it here, not to refactor everything at once.
We use a lot of `static` functions that are not 100% "static," IMO.

[_PHP spec_](https://www.php.net/manual/en/language.oop5.static.php)
> Declaring class properties or methods as static makes them accessible without needing an instantiation of the class.

- `public static function get_enabled_events` uses `self::get`
- `public static function load_opt_out` uses `self::get`
- `public static function get_product_identifier` uses `self::get`
- `public static function get` uses the `self::settings`
**but** we set the `self::settings` in the instance constructor https://github.com/woocommerce/woocommerce-google-analytics-integration/blob/cdb60c08a9cb62f351f5d324472ad62b6eef7105/includes/class-wc-google-gtag-js.php#L47:L47

So we could change `self::get( * )` to `$this->get()` or event `$this->settings[ * ]` (IMHO, `get` just blurs the picture here w/o much benefit).


Do you also suggest changing `tracker_function_name`, `DEVELOPER_ID`, and `get_consent_modes` to be instance methods/props?

Reply:
I'd suggest in this case to access the methods through the current instance, which will make it easier in the future to remove `static` from those function declarations. We could do that now but then code snippets / other plugins would have a little harder time getting the instance.

```suggestion
				esc_js( $this->get( 'ga_id' ) ),
				esc_js( $this->tracker_function_name() ),
				esc_js( self::DEVELOPER_ID ),
				json_encode( $this->get_consent_modes() ),
				json_encode( $this->get_site_tag_config() )
```

Generally the purpose of the `get` helper function would be to enforce both the parameter type for key and the return value. However it seems the return value for non_scalar values is broken. If you do something like this you'll get a fatal error:
```php
self::$settings['ga_id'] = [];
var_dump( $this->get( 'ga_id' ) );
```

Where we do see it working is for example if we do something like:
```php
self::$settings['ga_id'] = 1;
var_dump( $this->get( 'ga_id' ), self::$settings['ga_id'] );
```
The first value will be a string, but the second will be an int, so we will be less certain what kind of value we will be working with if we don't use the `get` method.

Reply:
Thanks :) I changed the reference in non-static methods to use `$this->` in https://github.com/woocommerce/woocommerce-google-analytics-integration/pull/398/commits/867f69657d7a1e48713ab49ab0cb56d594ede9d7 (plus [tweaked](https://github.com/woocommerce/woocommerce-google-analytics-integration/pull/398/commits/89487e643cb47a2030be2de9160d3b011532dd18) `self::DEVELOPER_ID` to `static::DEVELOPER_ID`, to suppress the notice).

Let's change the `static` in methods in a separate PR and release. I'll start a draft to keep this idea on track

Reply:
> Generally the purpose of the get helper function would be to enforce both the parameter type for key and the return value. However it seems the return value for non_scalar values is broken. If you do something like this, you'll get a fatal error:

I wonder if we really need such behavior. I was experimenting with adding array-like settings for consent modes, and this stringification behavior was more of a problem than a help. But that's also something to address in a separate PR

Reply:
Definitely limiting, and assuming of what settings are like. Agree that it would be better handled in a separate PR.

Reply:
Added an issue to track that enhancement idea https://github.com/woocommerce/woocommerce-google-analytics-integration/issues/412